### ISSUE TITLE
ExpandSignature Resolution Issues

### ISSUE THREAD

**Issue by chadhietala** [typed-ember/glint#405]
`ExpandSignature` is used within [@glint/template](https://github.com/typed-ember/glint/blob/main/packages/template/-private/index.d.ts#L9) which is pulling from [@glimmer/component@^1.1.2](https://github.com/typed-ember/glint/blob/main/packages/template/package.json#L21). Within `ExpandSignature` it uses [`ArgsFor`](https://github.com/glimmerjs/glimmer.js/blob/v1.x/packages/%40glimmer/component/addon/-private/component.ts#L48).

A problem arises with `ExpandSignature` when using `@glint/environment-glimmerx` because GlimmerX uses [@glimmer/component@2.0.0-beta.20](https://github.com/glimmerjs/glimmer-experimental/blob/master/packages/%40glimmerx/component/package.json#L18) and thus has this slightly different implementation of [`ArgsFor`](https://github.com/glimmerjs/glimmer.js/blob/master/packages/%40glimmer/component/addon/-private/component.ts#L54).

It surfaces as this error:

```
node_modules/@glint/template/-private/index.d.ts:53:5 - error TS2370: A rest parameter must be of an array type.

53     ...positional: ExpandSignature<S>['Args']['Positional']
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in node_modules/@glint/template/-private/index.d.ts:53
```

Playing around with the `ComponentLike` type in @glint/template  the following seems to resolve the issue but unsure if this is correct or not

```diff
export type ComponentLike<S = unknown> = InvokableConstructor<
  (
    named: GuardEmpty<ExpandSignature<S>['Args']['Named']>,
-    ...positional: ExpandSignature<S>['Args']['Positional']
+    ...positional: ExpandSignature<S>['Args']['Positional'] extends [] ? ExpandSignature<S>['Args']['Positional']  : []
  ) => AcceptsBlocks<
    FlattenBlockParams<ExpandSignature<S>['Blocks']>,
    ExpandSignature<S>['Element']
  >
>;
```

**Comment by dfreeman** [comment]
[RFC 748](https://github.com/emberjs/rfcs/pull/748) established the canonical fully-expanded form for a `@glimmer/component` signature, so I think if v2's `ExpandSignature` isn't consistent with that it probably represents a bug on that side.

Between @chriskrycho and me I think we ended up revising the v1 `ExpandSignature` implementation a couple of times to get everything working smoothly, and it may be that we never ported those revisions to the v2 PR that was still pending at that point.

Long story short, I think `ExpandSignature` should probably be identical in v1 and v2 of `@glimmer/component` since we have a documented specification for what the expanded signature is expected to look like.
